core/domain/exp_domain.py

```diff
@@ -2514,12 +2515,1063 @@ def _convert_states_v50_dict_to_v51_dict(
+        choices_to_remove = []
+        invalid_choices_index = []
+        for choice in choices:
+            if choice['html'].strip() in ('<p></p>', ''):
+                empty_choices.append(choice)
+
+        # Only one choice is empty.
+        if len(empty_choices) == 1:
+            invalid_choices_index.append(choices.index(empty_choices[0]))
+            choices.remove(empty_choices[0])
+        # Multiple choices are empty.
+        else:
+            for idx, empty_choice in enumerate(empty_choices):
+                empty_choice['html'] = (
+                    '<p>' + 'Choice ' + str(idx+1) + '</p>'
+                )
+
+        # Duplicate choices.
+        for choice in choices:
+            if choice['html'] not in seen_choices:
+                seen_choices.append(choice['html'])
+            else:
+                choices_to_remove.append(choice)
+                invalid_choices_index.append(choices.index(choice))
+
+        for choice_to_remove in choices_to_remove:
+            choices.remove(choice_to_remove)
+
+        # Remove rules whose choice has been deleted.
+        for answer_group in answer_groups:
+            invalid_rules = []
+            for rule_spec in answer_group['rule_specs']:
+                if rule_spec['rule_type'] == 'Equals':
+                    if rule_spec['inputs']['x'] in invalid_choices_index:
+                        invalid_rules.append(rule_spec)
+            for invalid_rule in invalid_rules:
+                answer_group['rule_specs'].remove(invalid_rule)
+            # Removal of answer group if it is valid check.
+
+        return (choices, answer_groups)
+
+    @classmethod
+    def _item_selec_no_ans_group_should_be_same(
+        cls, answer_groups: List[state_domain.AnswerGroup]
+    ) -> List[state_domain.AnswerGroup]:
+        """No answer group should be same in ItemSelectionInput interaction,
+        Specefically checks for `Equals` rule spec if there are similar
+        rules then removes the later one
+
+        Args:
+            answer_groups: List[state_domain.AnswerGroup]. The answer group.
+
+        Returns:
+            List[state_domain.AnswerGroup]. The modified answer group.
+        """
+        equals_rule_values = []
+        equals_rule_to_remove = []
+        for answer_group in answer_groups:
+            for rule_spec in answer_group['rule_specs']:
+                if rule_spec['rule_type'] == 'Equals':
+                    rule_spec_x = rule_spec['inputs']['x']
+                    if rule_spec_x in equals_rule_values:
+                        equals_rule_to_remove.append(rule_spec)
+                    else:
+                        equals_rule_values.append(rule_spec_x)
+
+        for rule_to_remove in equals_rule_to_remove:
+            for answer_group in answer_groups:
+                for rule_spec in answer_group['rule_specs']:
+                    if rule_spec == rule_to_remove:
+                        answer_group['rule_specs'].remove(rule_spec)
```

Comment:
I don't think this works... does it? Doesn't this end up removing all of the duplicate rule specs? Should it keep the 1st time the rule spec shows up? Also this triple `for` loop is not good -- try to modify all of your code to use Python `set` and `dict` wherever they may be appropriate.

Reply:
No it will not, the `equals_rule_to_remove` only contains the rule that needs to be removed as on line 2682 `if rule_spec_x in equals_rule_values:` we are checking if the value is present in `equals_rule_values` then only we append the invalid rule.

Yeah I agree that 3 loops are not very effective but the thing is we will not be able to use any other thing. The problem is that in python we cannot remove multiple items in the loop on which we are iterating, it will result in skipping some of the values. We cannot keep track of the indexes because as soon as we will remove one element the index will change. We will not be able to use the `set` as it is not necessary that the rule will be exactly same though they will be of rule_type `Equals` but some values like maybe destination node or solution part will differ, We still need to remove that as this rule will never going to get matched.
Thanks

Reply:
Okay! It will be faster if you make `for rule_to_remove in equals_rule_to_remove:` the 3rd for loop instead of the first.